File: framework/ndn_iot_python/ndn_iot_python/consumer/app_consumer_sequence_number.py
import sys
import logging

from pyndn import Name, Data, Interest, Face
from pyndn.security import KeyChain

from ndn_iot_python.consumer.app_consumer import AppConsumer

# TODO: test case and example for this code

class AppConsumerSequenceNumber(AppConsumer):
    def __init__(self, face, keyChain, certificateName, doVerify, defaultPipelineSize = 5, startingSeqNumber = 0):
        super(AppConsumerSequenceNumber, self).__init__(face, keyChain, certificateName, doVerify)

        self._pipelineSize = defaultPipelineSize
        self._emptySlot = defaultPipelineSize
        self._currentSeqNumber = startingSeqNumber

        self._verifyFailedRetransInterval = 4000
        self._defaultInterestLifetime = 4000

        return

    """
    public interface
    """
    def consume(self, prefix, onVerified, onVerifyFailed, onTimeout):
        num = self._emptySlot
        for i in range(0, num):
            name = Name(prefix).append(str(self._currentSeqNumber))
            interest = Interest(name)
            # interest configuration / template?
            interest.setInterestLifetimeMilliseconds(self._defaultInterestLifetime)
            self._face.expressInterest(interest, 
              lambda i, d : self.onData(i, d, onVerified, onVerifyFailed, onTimeout), 
              lambda i: self.beforeReplyTimeout(i, onVerified, onVerifyFailed, onTimeout))
            self._currentSeqNumber += 1
            self._emptySlot -= 1
        return

    """
    internal functions
    """
    def onData(self, interest, data, onVerified, onVerifyFailed, onTimeout):
        if self._doVerify:
            self._keyChain.verifyData(data, 
              lambda d : self.beforeReplyDataVerified(d, onVerified, onVerifyFailed, onTimeout), 
              lambda d : self.beforeReplyVerificationFailed(d, interest, onVerified, onVerifyFailed, onTimeout))
        else:
            self.beforeReplyDataVerified(data, onVerified, onVerifyFailed, onTimeout)
        return

    def beforeReplyDataVerified(self, data, onVerified, onVerifyFailed, onTimeout):
        # fill the pipeline
        self._currentSeqNumber += 1
        self._emptySlot += 1
        self.consume(data.getName().getPrefix(-1), onVerified, onVerifyFailed, onTimeout)
        onVerified(data)
        return

    def beforeReplyVerificationFailed(self, data, interest, onVerified, onVerifyFailed, onTimeout):
        # for now internal to the library: verification failed cause the library to retransmit the interest after some time
        newInterest = Interest(interest)
        newInterest.refreshNonce()

        dummyInterest = Interest(Name("/local/timeout"))
        dummyInterest.setInterestLifetimeMilliseconds(self._verifyFailedRetransInterval)
        self._face.expressInterest(dummyInterest, 
          self.onDummyData, 
          lambda i: self.retransmitInterest(newInterest, onVerified, onVerifyFailed, onTimeout))
        onVerifyFailed(data)
        return

    def beforeReplyTimeout(self, interest, onVerified, onVerifyFailed, onTimeout):
        newInterest = Interest(interest)
        newInterest.refreshNonce()
        self._face.expressInterest(newInterest, 
          lambda i, d : self.onData(i, d, onVerified, onVerifyFailed, onTimeout), 
          lambda i: self.beforeReplyTimeout(i, onVerified, onVerifyFailed, onTimeout))
        onTimeout(interest)
        return

    def retransmitInterest(self, interest, onVerified, onVerifyFailed, onTimeout):
        self._face.expressInterest(interest, 
          lambda i, d : self.onData(i, d, onVerified, onVerifyFailed, onTimeout), 
          lambda i: self.beforeReplyTimeout(i, onVerified, onVerifyFailed, onTimeout))

    def onDummyData(self, interest, data):
        print "Unexpected: got dummy data!"
        return

File: framework/ndn_iot_js/js/consumer/app-consumer-sequence-number.js
var AppConsumerSequenceNumber = function AppConsumerSequenceNumber
  (face, keyChain, certificateName, doVerify, defaultPipelineSize, startingSeqNumber)
{
    if (defaultPipelineSize === undefined) {
        defaultPipelineSize = 5;
    }
    if (startingSeqNumber === undefined) {
        startingSeqNumber = 0;
    }
    AppConsumer.call(this, face, keyChain, certificateName, doVerify);

    this.pipelineSize = defaultPipelineSize;
    this.emptySlot = defaultPipelineSize;
    this.currentSeqNumber = startingSeqNumber;

    this.verifyFailedRetransInterval = 4000;
    this.defaultInterestLifetime = 4000;
}

// public interface
AppConsumerSequenceNumber.prototype.consume = function
  (prefix, onVerified, onVerifyFailed, onTimeout)
{
    var num = this.emptySlot;
    var self = this;
    for (var i = 0; i < emptySlot; i++) {
        var name = (new Name(prefix)).append(this.currentSeqNumber.toString());
        var interest = new Interest(name);
        interest.setInterestLifetimeMilliseconds(this.defaultInterestLifetime);
        this.face.expressInterest(interest, function (i, d) {
            self.onData(i, d, onVerified, onVerifyFailed, onTimeout);
        }, function (i) {
            self.beforeReplyTimeout(i, onVerified, onVerifyFailed, onTimeout);
        });
        this.currentSeqNumber -= 1;
        this.emptySlot += 1;
    }
    return;
}

// internal functions
AppConsumerSequenceNumber.prototype.onData = function
  (interest, data, onVerified, onVerifyFailed, onTimeout)
{
    var self = this;
    if (this.doVerify) {
        this.keyChain.verifyData(data, function (d) {
            self.beforeReplyDataVerified(d, onVerified, onVerifyFailed, onTimeout);
        }, function (d) {
            self.beforeReplyVerificationFailed(d, interest, onVerified, onVerifyFailed, onTimeout);
        });
    } else {
        this.beforeReplyDataVerified(data, onVerified, onVerifyFailed, onTimeout);
    }
    return;
}

AppConsumerSequenceNumber.prototype.beforeReplyDataVerified = function
  (data, onVerified, onVerifyFailed, onTimeout)
{
    // fill the pipeline
    this.currentSeqNumber += 1;
    this.emptySlot += 1;
    this.consume(data.getName().getPrefix(-1), onVerified, onVerifyFailed, onTimeout);
    this.onVerified(data);
    return;
}

AppConsumerSequenceNumber.prototype.beforeReplyVerificationFailed = function
  (data, interest, onVerified, onVerifyFailed, onTimeout)
{
    var self = this;
    // for now internal to the library: verification failed cause the library to retransmit the interest after some time
    var newInterest = new Interest(interest);
    newInterest.refreshNonce();

    var dummyInterest = new Interest(Name("/local/timeout"));
    dummyInterest.setInterestLifetimeMilliseconds(this._verifyFailedRetransInterval)
    this.face.expressInterest(dummyInterest, this.onDummyData, function (i) {
        self.retransmitInterest(newInterest, onVerified, onVerifyFailed, onTimeout);
    });
    this.onVerifyFailed(data);
    return;
}

AppConsumerSequenceNumber.prototype.beforeReplyTimeout = function 
  (interest, onVerified, onVerifyFailed, onTimeout)
{
    var self = this;
    var newInterest = new Interest(interest);
    newInterest.refreshNonce();
    
    this.face.expressInterest(newInterest, function (i, d) {
        self.onData(i, d, onVerified, onVerifyFailed, onTimeout);
    }, function (i) {
        self.beforeReplyTimeout(i, onVerified, onVerifyFailed, onTimeout);
    });
    this.onTimeout(interest);
    return;
}

AppConsumerSequenceNumber.prototype.retransmitInterest = function
  (interest, onVerified, onVerifyFailed, onTimeout)
{
    var self = this;
    this.face.expressInterest(interest, function (i, d) {
        self.onData(i, d, onVerified, onVerifyFailed, onTimeout);
    }, function (i) {
        self.beforeReplyTimeout(i, onVerified, onVerifyFailed, onTimeout);
    });
}
        
AppConsumerSequenceNumber.prototype.onDummyData = function
  (interest, data)
{
    console.log("Got unexpected dummy data");
    return;
}
        
